### PR TITLE
Add arbitrary tensor output during training/testing

### PR DESCRIPTION
**Pull request by angeloskath**
This PR is in reference to issue #6414. The purpose is to provide with a view under the hood for people who want to evaluate what is happening during training. I will showcase in the following code snippet what is proposed but the main goal is to create a discussion because nobody seemed interested to counter-argue with the issue.

```python
# Let's say I want to count how many values are negative before relu.
# It is currently not possible in keras during training. We could use a metric but we
# would not be able to access the values and we would have to add an output
# not used during prediction.
from keras.layers import Activation, Dense, Dropout, Input
from keras.models import Model
import numpy as np

x0 = Input(shape=(10,))
x = Dense(10, activation="relu")(x0)
x = Dropout(0.5)(x)
y_extra = x = Dense(10)(x)
x = Activation("relu")(x)
x = Dropout(0.5)(x)
y = Dense(1)

m = Model(inputs=[x0], outputs=[y], extra_output=[y_extra])
m.compile(optimizer="sgd", loss="mse")

loss, y_extra = m.train_on_batch(np.random.rand(128, 10), np.random.rand(128, 1))
loss = m.evaluate(np.random.rand(128, 10), np.random.rand(128), batch_size=10)
y = m.predict(np.random.rand(128, 10), batch_size=10)
```

So, I am waiting to hear your thoughts on the matter.

### CODE REVIEW

**Comment by Dref360** [comment]
Why not just use K.function in this case? 
Great feature imo but I guess there is already a way to do that with K.function. your wouldn't get the callbacks called

**Comment by angeloskath** [comment]
Using `K.function`, you would need to reimplement all the code in Keras that uses the optimizer and the loss to create the train function.

@fchollet, what do you think?

**Comment by fchollet** [comment]
Keras models are already graphs. I don't see a need for "extra outputs", these are just outputs. The outputs of a Keras model are defined at instantiation time via the `outputs` argument.

Note that not all outputs of a Keras model have to have a loss associated with them. You can pass `None` as loss for any output, in which it won't be used for training and no target data will be expected for it during `fit`/ etc.

**Comment by angeloskath** [comment]
Extra is not the correct name. The main difference is that there is no way
to get an output during training. train_on_batch() simply returns scalars.

So the main question​ is how would you look under the hood during training?
How would you get the output of the n-th layer during training? How would
you get the number of non-zero values after a RELU layer during training?
How would you get the per sample loss during training?

Extra output is just a bad name which should change if the PR is to be
merged (open to suggestions by the way).

To sum up, I know how to create a model that has arbitrary outputs. The
issue is with arbitrary outputs during training. If you look closely to the
PR it only changes the train and test functions, not the predict function.

I agree that the API is not necessarily good. I am open to suggestions
regarding how should we implement arbitrary outputs during training.

On May 27, 2017 2:01 AM, "François Chollet" <notifications@github.com>
wrote:

Keras models are already graphs. I don't see a need for "extra outputs",
these are just outputs. The outputs of a Keras model are defined at
instantiation time via the outputs argument.

Note that not all outputs of a Keras model have to have a loss associated
with them. You can pass None as loss for any output, in which it won't be
used for training and no target data will be expected for it during fit/
etc.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/fchollet/keras/pull/6733#issuecomment-304411363>, or mute
the thread

**Comment by fchollet** [comment]
> So the main question​ is how would you look under the hood during training?

`K.function` inside a callback. If you need total customization, you can also use Keras to build your model then use pure TensorFlow to manipulate it and train it (if you are using `train_on_batch`, you are already nearly at that stage, so it isn't much added work to drop the Keras training routines entirely).
